### PR TITLE
Parsing of complex tiles such as those in Thai and Lao scripts

### PR DESCRIPTION
**Pull request by laineyhm**
In this PR,
1) Methods were added and changed to allow parsing of complex tiles such as those in Thai and Lao script languages
2) A readability improvement was made, from importing inline e.g. Start.tileList, Start.wordList, etc, to importing once at the top and then referring to these static objects simply as tileList, wordList, etc.

Since tile type is necessary information for parsing complex tiles, several methods have new parameters used for accessing this type information, e.g. a reference to the original word when parsing a pseudoword, in order to access the wordlist mixed types info.

There are two levels of word parsing. Both return the same thing in strictly unidirectional parsing, and they differ as follows for Thai and Lao script parsing: 
parseWordIntoTilesPreliminary() is a greedy search that goes from beginning to end of the word, getting the longest sequences possible that match tiles in the tile list (which can be broken up with periods). 
If the script type given in langinfo is "Thai" or "Lao", parseWordIntoTiles() then goes over this preliminary tile array and finds any complex tiles to combine, using type information and script structure. The default, or with any other script type given in langinfo, is to return the preliminaryTileArray from parseWordIntoTiles(). 
So, parseWordIntoTiles (a method of the tileList) is the authoritative source for the tiles in a word. But, the preliminaryTileArray is used in wordsearch and a few other games, and for accessing tile type info from the word list.

### CODE REVIEW

**Review by nathanchou21 — COMMENTED** [review]
I don't have many comments or anything, but reading through the parsing and mixed type methods was definitely useful. I'll try to copy and paste as much over as I can, though working without all the start fields and classes that are used might get tricky. Let me know if you have more suggestions!

File: app/src/main/java/org/alphatilesapps/alphatiles/Start.java [hunk 1]
```diff
@@ -1548,6 +1608,235 @@ public ArrayList<String> parseWordIntoTiles(String parseMe) {
             return parsedWordArrayTemp;
         }
 
+        /*
+        LM: parseWordIntoTiles does not currently take into consideration symbols of type X because we don't have
+        data to show how these tiles pattern in Thai script language packs yet. 03-2023
+         */
+        public ArrayList<String> parseWordIntoTiles(String stringToParse, String wordInLOP) {
```
Comment (nathanchou21) [inline app/src/main/java/org/alphatilesapps/alphatiles/Start.java:1615]: Would you mind explaining what the stringToParse parameter is for? At least based on command F it seems like the method is always called with the same value passed in for the two parameters, with stringToParse only being used once. Thanks!

File: app/src/main/java/org/alphatilesapps/alphatiles/Start.java [hunk 2]
```diff
@@ -1548,6 +1608,235 @@ public ArrayList<String> parseWordIntoTiles(String parseMe) {
...
+                        }
+                    } else {
+                        currentTileType = currentTile.tileType;
+                    }
+                    if (currentTileType.equals("AV") || currentTileType.equals("BV") || currentTileType.equals("FV")) { // Prepare to add current AV/BV/FV to vowel-so-far
+                        vowelTile = tileHashMap.get(vowel);
+                        if(!Objects.isNull(vowelTile)){ // Vowel composite so far is parsable as one tile in the tile list
+                            vowelTileString = vowelTile.baseTile;
+                            if (MULTIFUNCTIONS.contains(vowelTileString)) { // Discern type of current vowel composite
+                                int indexInPreliminaryArray = returnInstanceIndexInPreliminaryParsedWordArray(vowelTileString, indexOfLastFoundMultitypeVowel, parsedWordArrayPreliminary, wordInLOP);
```
Comment (nathanchou21) [inline app/src/main/java/org/alphatilesapps/alphatiles/Start.java:1727]: FYI, if you haven't already, you may need to make your own test case for discerning a composite. When I put a break point at line 1727, it didn't trigger for any of the two Thai or the Lao language packs.

**Comment by laineyhm** [comment]
> I don't have many comments or anything, but reading through the parsing and mixed type methods was definitely useful. I'll try to copy and paste as much over as I can, though working without all the start fields and classes that are used might get tricky. Let me know if you have more suggestions!

Yes, changing the imports in the same PR was poor form. I should have made a separate PR to main with that, and then rebased this PR on top of that. What I did instead was just make the change for myself in my branch because I preferred it...but then we have to deal with differences in main. Sorry about that! I learned not to do that I think!

File: app/src/main/java/org/alphatilesapps/alphatiles/Start.java [hunk 3]
```diff
@@ -233,9 +240,9 @@ protected void onCreate(Bundle savedInstanceState) {
         }
 
         if (differentiateTypes && MULTIFUNCTIONS.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (laineyhm) [reply app/src/main/java/org/alphatilesapps/alphatiles/Start.java]: Good idea. Done

File: app/src/main/java/org/alphatilesapps/alphatiles/Start.java [hunk 4]
```diff
@@ -331,7 +338,7 @@ public void buildTilesArray() {
                 Tile tile = new Tile(thisLineArray[0], thisLineArray[1], thisLineArray[2], thisLineArray[3], thisLineArray[4], thisLineArray[5], thisLineArray[6], thisLineArray[7], thisLineArray[8], thisLineArray[9], thisLineArray[10], 0, 0, 0, stageOfFirstAppearance, stageOfFirstAppearanceType2, stageOfFirstAppearanceType3);
                 if (!tile.hasNull()) {
                     tileList.add(tile);
-                    if (!tile.tileType.equals("SAD")) {
+                    if (!tile.tileType.equals("SAD") && !tile.audioForTile.equals("zz_no_audio_needed")) {
```
[the remark this answers is not in the document]
Comment (laineyhm) [reply app/src/main/java/org/alphatilesapps/alphatiles/Start.java:340]: Yes, that is correct. Right now, the default for language packs that don't have tile audio is to put "X" for audio name. zz_no_audio_needed is for pieces of tiles that don't occur on their own, only as components of complex tiles. Currently, we need these "tile pieces" to appear in the word list because we need their type info to know how they combine with other parts of the complex tiles. But they don't have their own audio.

File: app/src/main/java/org/alphatilesapps/alphatiles/Start.java [hunk 5]
```diff
@@ -348,11 +355,11 @@ public void buildTilesArray() {
             for (Tile tile : tileList) {
                 tileListWithMultipleTypes.add(tile.baseTile);
                 tileTypeHashMapWithMultipleTypes.put(tile.baseTile, tile.tileType);
-                if (!tile.tileType.equals("SAD")) {
+                if (!tile.tileType.equals("SAD") && !tile.audioForTile.equals("zz_no_audio_needed")) { // From SAD, also exclude preliminary e.g. Thai script "tiles" that combine to form whole tiles but don't appear alone in the language
```
[the remark this answers is not in the document]
Comment (laineyhm) [reply app/src/main/java/org/alphatilesapps/alphatiles/Start.java]: Yes, thank you!

File: app/src/main/java/org/alphatilesapps/alphatiles/Start.java [hunk 6]
```diff
@@ -30,6 +33,8 @@ public class Start extends AppCompatActivity {
 
     public static String localAppName; // KP add "public"
 
+    public static String scriptType; // LM Can be "Thai". If nothing specified, tile parsing defaults to unidirectional.
```
[the remark this answers is not in the document]
Comment (laineyhm) [reply app/src/main/java/org/alphatilesapps/alphatiles/Start.java]: Yes, thank you!

File: app/src/main/java/org/alphatilesapps/alphatiles/Start.java (hunk 1 above)
Comment (laineyhm) [reply app/src/main/java/org/alphatilesapps/alphatiles/Start.java:1618]: I just added a new comment to the method header about those parameters. Let me know if you have any further questions. 
TLDR: Sometimes the stringToParse isn't something in the wordlist, but rather a pseudoword for an incorrect answer choice. But we need real word information to know the correct way to parse Thai/Lao tiles.

File: app/src/main/java/org/alphatilesapps/alphatiles/Start.java [hunk 7]
```diff
@@ -1548,6 +1608,235 @@ public ArrayList<String> parseWordIntoTiles(String parseMe) {
...
+                foundNextConsonant = false;
+                Tile currentTile;
+                String currentTileString;
+                String currentTileType;
+                // Scan for the next unchecked consonant tile
+                while (!foundNextConsonant && consonantScanIndex < parsedWordArrayPreliminary.size()) {
+                    currentTileString = parsedWordArrayPreliminary.get(consonantScanIndex);
+                    currentTile = tileHashMap.get(currentTileString);
+                    if (MULTIFUNCTIONS.contains(currentTileString)) { // Discern the type of the current tile
+                        int indexInPreliminaryArray = returnInstanceIndexInPreliminaryParsedWordArray(currentTileString, consonantScanIndex, parsedWordArrayPreliminary, wordInLOP);
```
[the remark this answers is not in the document]
Comment (laineyhm) [reply app/src/main/java/org/alphatilesapps/alphatiles/Start.java:1647]: I will look into this, as well as your previous comment. I agree that the current structure is cumbersome. Changing to using Tile lists rather than String lists will require refactors in all of the game classes for accessing the Strings as properties of Tiles instead of directly, so I will look at the tradeoffs.
